### PR TITLE
validation: port some of the functionality from reana-client

### PR DESCRIPTION
**Pull request by audrium**
closes https://github.com/reanahub/reana-server/issues/442

TODO:
- [ ] Double test all the edge cases 
- [ ] Add changelog message
- [ ] Port and fix tests

### CODE REVIEW

File: reana_commons/config.py [hunk 1]
```diff
@@ -350,14 +356,12 @@ def workspaces(paths):
 WORKSPACE_PATHS = workspaces(json.loads(os.getenv("WORKSPACE_PATHS", "{}")))
 """Dictionary of available workspace paths with pairs of cluster_node_path:cluster_pod_mountpath."""
 
+AVAILABLE_WORKSPACE_PATHS = list(WORKSPACE_PATHS.values())
```
Comment (tiborsimko) [inline reana_commons/config.py:359]: I think it may be confusing to have both `WORKSPACE_PATHS` and `AVAILABLE_WORKSPACE_PATHS` as two different configuration variables for denoting the same thing.

Perhaps you could keep only `WORKSPACE_PATHS` and do the `list()` in each use case occurrence, or else introduce a convenience function `get_available_workspace_paths()` that would return list that could be directly consumed, and `get_default_workspace()` to return default one, etc. Then one-config, several-convenience-functions might result in more readable code.

P.S. That said, when I see `WORKSPACE_PATHS`, my first reaction is to think about a list, not a dictionary... So we may have a slight misnomer there too.

File: reana_commons/validation/compute_backends.py [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+
+    def raise_error(self, compute_backend: str, step_name: str) -> None:
+        """Raise validation error."""
+        raise REANAValidationError(
+            f'Compute backend "{compute_backend}" found in step "{step_name}" is not supported. '
+            f'List of supported compute backends: "{", ".join(self.supported_backends)}"'
+        )
+
+
+class SerialComputeBackendValidator(ComputeBackendValidatorBase):
```
Comment (tiborsimko) [inline reana_commons/validation/compute_backends.py:76]: Nomenclature comment: if the abstract class is `ComputeBackendValidator`, then specific classes are better named `ComputeBackendValidatorSerial`, `ComputeBackendValidatorYadage`, etc.  This is usually what we use elsewhere, e.g. `JobMonitor` and `JobMonitorKubernetes`, `JobMonitorHTCondor`, etc.  Advantage being the common prefix makes it easy to identify related classes when grepping etc.  

(Kind of like `module.foo`, `module.bar` name association idea.)

File: reana_commons/validation/parameters.py [hunk 3]
```diff
@@ -0,0 +1,454 @@
...
+                            type=workflow_type.capitalize(),
+                            parameter=parameter,
+                            steps=", ".join(steps_diff),
+                            s="s" if len(steps_diff) > 1 else "",
+                        ),
+                    }
+                )
+
+
+class SerialParameterValidator(ParameterValidatorBase):
```
Comment (tiborsimko) [inline reana_commons/validation/parameters.py:188]: The same naming comment here, `ParameterValidator` prefix would call for `ParameterValidatorSerial`, `ParameterValidatorYadage`, etc.

File: reana_commons/validation/compute_backends.py (hunk 2 above)
Comment (tiborsimko) [reply reana_commons/validation/compute_backends.py:76]: (Note: We can leave this one for later, as the original code classes were also named like that.)

Comment (audrium) [reply reana_commons/validation/compute_backends.py]: Thanks, changed the naming for all validation classes accordingly

File: reana_commons/config.py (hunk 1 above)
Comment (audrium) [reply reana_commons/config.py]: Indeed, reverted to the previous logic where we had only `WORKSPACE_PATHS` and used `list()` in each of the use case as suggested

**Review by tiborsimko — APPROVED** [review]
